Modify Q.eq

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Related to #20965

#### Brief description of what is fixed or changed

This change makes `Q.eq` completely internal to assumptions module. Please refer the the discussion in  https://groups.google.com/g/sympy/c/jOwmNkN2hF4

#### Other comments

Some bugs have been found and resolved during the process.

1. `refine()` on `Boolean` now returns `true` and `false` instead of `True` and `False`.
2. `sympy_parse` module can now parse applied predicates

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * `Q.is_true` wrapping over `AppliedPredicate` now just return the argument.
<!-- END RELEASE NOTES -->

JSS95 commented:
This change just reorganizes the code, so if there is no review I will merge this change. Please note that the general idea was discussed and approved in #20929 and #20965. I plan to add other predicates such as `Q.ge`, `Q.lt`, etc and write the detailed documentation in succeeding PRs.

oscarbenjamin commented on sympy/printing/mathml.py:
```diff
@@ -536,6 +536,13 @@ def _print_Relational(self, e):
         x.appendChild(self._print(e.rhs))
         return x
 
+    def _print_AppliedBinaryRelation(self, e):
```
Since this is just a duplicate of the method above I think it would make more sense to have both methods call a helper.

JSS95 replied:
Thanks. I found out that this method had problem so I enhanced it and added tests.

oscarbenjamin commented on sympy/printing/julia.py:
```diff
@@ -456,6 +456,21 @@ def _print_Piecewise(self, expr):
                     lines.append("end")
             return "\n".join(lines)
 
+    def _print_AppliedBinaryRelation(self, expr):
```
Is there any reason this should print differently from `Relational`?

oscarbenjamin commented:
Why do we need to be able to use `Q.eq` etc with the code printers. We have `Eq` which is supposed to evaluate to Boolean so that makes sense. I would have thought that in any situation where we wanted a code printer we should probably be using `Relational` rather than `AppliedPredicate`.

JSS95 commented:
Oh, that's right. We don't need it.

JSS95 commented:
I removed the code printing of `Q.eq`.

asmeurer commented:
What is the final goal with all of this? If we ever add Eqn (Equation), there will be three distinct equality object, Eq, Eqn, and Q.eq. Why do we need three of them? Eq and Q.eq be aliases for the same thing? If so, can we make this change without introducing a new public name in the intermediate? If the main concern is that Eq evaluates sometimes, can we just remove that (maybe keep it under `doit()`)? Another concern is that users are already using Eq, but I think Oscar already argued that for Eqn to really work, users are going to have to migrate to it deliberately (IIRC).

JSS95 commented:
@asmeurer 
The goal is to make relation as predicate so that we use `Q.eq(x, y)` instead of `Q.is_true(Eq(x, y))`, and allow inference of relational such as `Q.gt(x, y) -> AllArgs(Q.real)`. We can aliase `Eq` to return `Q.eq`, but not vice versa. `Eq` is quite widely used around SymPy so I plan to introduce the relational predicate first and make aliasing later.

asmeurer commented:
Why not the other way around? If we want Eq and Q.eq to be the same, why do we need a new name Q.eq? We can just use Eq. Predicate objects don't have to be on Q. It seems unnecessary and confusing to have two names for the same object, if that is the ultimate goal.

JSS95 commented:
It can be done. So, your suggestion is to make `Eq` as a predicate object and make `Eq(x, x)` not evaluate to `True` but return `AppliedBinaryRelation` instance, right?

oscarbenjamin commented:
I'm not sure I understand what you are suggesting. It really would be better if you would write up your overall idea somewhere (for Predicate, Relational, how it relates to Equation etc). At the moment there is discussion split over many long PRs some of which are merged and some of which are closed. Your idea seems to have changed multiple times and I don't think anyone else knows what it is you are trying to achieve.

> make `Eq` as a predicate object and to make `Eq(x, x)` not evaluate to `true`

That is not a backwards compatible change if I understand what you mean correctly. The fact that `Eq(x, y)` tests for equality and evaluates to true is depended on in the sympy codebase and also by downstream users.

JSS95 commented:
My idea is to ***temporarily*** have two separate systems to represent a relation : `Eq`, `Gt`, etc and `Q.eq`, `Q.gt`, etc. The naming is tentative - I do not assert to register them on `Q`.

The reason for having two systems is to ensure backwards compatibility because `Eq` and `Q.eq` have different behavior. After some release period, `Q.eq` can fully replace - and be aliased as - `Eq`.

This is my overall idea about relational predicate :

### evaluation

`Q.eq` is a predicate, and `Q.eq(x, y)` returns applied predicate. Like any other applied predicate, `Q.eq(x, y)` is never evaluated by construction or simplification. In other words, `Q.eq(x, x)`, `Q.eq(x, x).doit()` or `Q.eq(x, x).simplify()` does not return `true`. Instead, `ask(Q.eq(x, x))` returns `True` and `refine(Q.eq(x, x))` returns `S.true`.

### symbolic manipulation

Since `Q.eq(x, y)` is never automatically evaluated, we can safely introduce symbolic manipulation of the relation.

`Q.eq(x, y)` is strictly boolean and does not have direct symbolic manipulation methods. For example `Eq(x, y).expand()` returns `Eq(x.expand(), y.expand())`, but `Q.eq(x, y).expand()` is not allowed.

Instead, `Q.eq(x, y)` will have proxy object as @oscarbenjamin suggested for 'Equation' in [this comment](https://github.com/sympy/sympy/pull/19479#issuecomment-637859457), which can be called by `.apply` property. `Q.eq(x, y).apply.expand()` will return `Q.eq(x.expand(), y.expand())`.

Also, `Q.eq(x, y)` will be able to support algebra between the relational as @sylee957 shared in [this comment](https://github.com/sympy/sympy/pull/20723#issuecomment-763975854). `Q.eq(x, 1) + Q.eq(y, 1)` will return `AddSides(Q.eq(x, 1), Q.eq(y, 1), evaluate=True)`, which in turn will be evaluated to `Q.eq(x + y, 2)`. `Q.gt(x, 2) * -1` will return `Q.lt(-x, -2)`, and `Q.gt(x, 2) * y` will return `MultiplySides(Q.gt(x, 2), y)` if the sign of `y` cannot be determined.

This way, `Q.eq` can make `Equation` no longer be required to be implemented. However I am willing to discard this idea if this does not get approved.

### compatibility with Eq around codebase

I will modify the codebase so that `Q.eq` can be used instead of `Eq` everywhere. For example `Piecewise((1, Q.eq(x, 0)), (0, True))` will be supported, and `solveset(Q.eq(x**2, 1))` will return `{-1, 1}`.

### assumption

Unlike `Eq` which needs to be wrapped by `Q.is_true`, `Q.eq(x, y)` itself can be valid proposition and assumption.

Relational predicates will be compatible with inference and SAT algorithms. For example, `Q.gt(x, y)` will imply that `x` and `y` are both comparable so that `ask(Q.extended_real(x), Q.gt(x, y))` returns `True`.

### migration from `Eq` to `Q.eq`

`Eq` and other relational will be retained only for backwards compatibility. Documentation for `Q.eq` will be thoroughly made and users will be informed to use `Q.eq` instead of `Eq`. After a few release, `Q.eq` will be renamed to `Eq`.

JSS95 commented:
And, if I understand correctly, @asmeurer 's idea is to skip the migration period and to name the equality predicate as `Eq` immediately. Am I correct? I do not disagree with this, but I still prefer to have two relational system for the time being.

oscarbenjamin commented:
I think that the plan you are writing should be written somewhere that is not in a pull request (e.g. on the wiki). Since the other Equation proposal has been discussed on the mailing list then this should be as well. Especially if the suggestion is to break backwards compatibility for Eq at some point then it needs to be discussed more widely as Eq is widely used.

> If this sounds strange, please refer to the solvers module where `solveset` is planned to replace (and eventually be renamed as) `solve` : https://docs.sympy.org/latest/modules/solvers/solveset.html#what-will-you-do-with-the-old-solve

Those docs should be fixed because I don't think we can ever rename `solveset` to `solve`. They are completely incompatible and `solve` is an extremely widely used function. I think it would be reasonable to say that we hope to use `solveset` in the implementation of `solve`. Potentially there could be some backwards incompatible changes to `solve` to remove some of the useless "features" that it has but the core signature that returns a list or list of dict etc can not be changed.

JSS95 commented:
Okay, I made a wiki page https://github.com/sympy/sympy/wiki/Relation-predicate

So, may I have your opinion about the general idea? We can have the discussion here and I will update the wiki page accordingly.

oscarbenjamin commented:
I think that this needs to be discussed more widely so you should post your suggestion to the mailing list.

It needs to be written though so that users can understand at least what exactly is being changed and how it would impact them. A justification should be given for breaking backwards compatibility and an explanation of what exactly would break and how users should mitigate for that.

JSS95 committed:
feat(assumptions) : Modify Q.eq and Q.ne

Remove the symbolic methods of binary predicates
Make Q.is_true on another applied predicate return the argument
Modify the printing of applied binary relations
Fix bugs in logic/boolalg
Fix bug in sympy_parser.py

JSS95 committed:
feat(assumptions) : Modify Q.eq and Q.ne

Modify the docstring
Make dispatching via relational predicate not possible

JSS95 committed:
feat(assumptions) : Modify Q.eq and Q.ne

Fix code quality

JSS95 committed:
feat(assumptions) : Modify Q.eq and Q.ne

Modify docstring of UnequalityPredicate
